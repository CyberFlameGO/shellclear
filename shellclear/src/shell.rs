use anyhow::anyhow;
use anyhow::Result;
use std::fmt;
use std::path::Path;
use strum::{EnumIter, IntoEnumIterator};

/// List of all supported shells
#[derive(Debug, EnumIter, Clone)]
pub enum Shell {
    Bash,
    Zshrc,
    Fish,
}

impl fmt::Display for Shell {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Self::Bash => write!(f, "bash"),
            Self::Zshrc => write!(f, "zshrc"),
            Self::Fish => write!(f, "fish"),
        }
    }
}

/// Zsh history file name
const ZSH_HISTORY_FILE_PATH: &str = ".zsh_history";
/// Bash history file name
const BASH_HISTORY_FILE_PATH: &str = ".bash_history";
const FISH_HISTORY_FILE_PATH: &str = ".local/share/fish/fish_history";

/// History shell details
#[derive(Clone, Debug)]
pub struct History {
    /// shell type
    pub shell: Shell,
    /// history file path
    pub path: String,
    /// history file name
    pub file_name: String,
}

/// return list of all existing history files
pub fn get_all_history_files() -> Result<Vec<History>> {
    let homedir = match dirs::home_dir() {
        Some(h) => h.display().to_string(),
        None => return Err(anyhow!("could not get directory path")),
    };

    // return list of existing shells
    Ok(Shell::iter()
        .map(|shell| {
            let shell_history_path = get_shell_history_path(&shell, &homedir);
            if !Path::new(&shell_history_path).exists() {
                log::debug!("shell {:?} not found", shell);
                return Err(anyhow!("could not get directory path"));
            }
            let file_name = Path::new(&shell_history_path)
                .file_name()
                .unwrap()
                .to_str()
                .unwrap()
                .to_string();
            Ok(History {
                shell,
                path: shell_history_path,
                file_name,
            })
        })
        .filter(std::result::Result::is_ok)
        .map(std::result::Result::unwrap)
        .collect::<Vec<_>>())
}

// returns all supported shells types
fn get_shell_history_path(shell_type: &Shell, homedir: &str) -> String {
    match shell_type {
        Shell::Bash => Path::new(homedir)
            .join(BASH_HISTORY_FILE_PATH)
            .display()
            .to_string(),
        Shell::Zshrc => Path::new(homedir)
            .join(ZSH_HISTORY_FILE_PATH)
            .display()
            .to_string(),
        Shell::Fish => Path::new(homedir)
            .join(FISH_HISTORY_FILE_PATH)
            .display()
            .to_string(),
    }
}
