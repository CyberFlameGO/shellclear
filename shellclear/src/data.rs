use crate::shell::Shell;
use serde_derive::Deserialize;

pub struct CmdExit {
    pub code: exitcode::ExitCode,
    pub message: Option<String>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct SensitiveCommands {
    #[serde(with = "serde_regex")]
    pub test: regex::Regex,
    pub name: String,
}

#[derive(Debug)]
pub struct FindingSensitiveCommands {
    pub shell_type: Shell,
    pub sensitive_findings: Vec<SensitiveCommands>,
    pub command: String,
}
