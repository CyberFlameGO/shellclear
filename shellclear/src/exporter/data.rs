use crate::data::FindingSensitiveCommands;
use crate::shell::Shell;
use anyhow::Result;
use chrono::{DateTime, Local, NaiveDateTime, TimeZone};
use lazy_static::lazy_static;
use regex::Regex;

const DATE_TIME_FORMAT: &str = "%Y-%m-%d %H:%M:%S";

pub trait Exporter {
    fn sensitive_data(&self, findings: &[&FindingSensitiveCommands]) -> Result<()>;
}

lazy_static! {
    static ref ZSHRC_CAPTURE_TIME: Regex = Regex::new(r"^: ([0-9]+):").unwrap();
}

// todo:: move under export trait
pub fn chunk(text: &str, size: usize) -> String {
    text.chars()
        .collect::<Vec<char>>()
        .chunks(size)
        .map(|c| c.iter().collect::<String>())
        .collect::<Vec<String>>()
        .join("\r\n")
}

/// Get command time execution
///
/// # Errors
///
/// Will return `Err` when conversion error
pub fn extract_time(finding: &FindingSensitiveCommands) -> Result<String> {
    match finding.shell_type {
        Shell::Zshrc => {
            if let Some(c) = ZSHRC_CAPTURE_TIME.captures(&finding.data) {
                if let Some(timestemp) = c.get(1) {
                    let date_time: DateTime<Local> = Local.from_utc_datetime(
                        &NaiveDateTime::from_timestamp(timestemp.as_str().parse::<i64>()?, 0),
                    );
                    return Ok(format!("{}", date_time.format(DATE_TIME_FORMAT)));
                }
            };
            Ok(finding.data.clone())
        }
        _ => Ok("".to_string()),
    }
}
