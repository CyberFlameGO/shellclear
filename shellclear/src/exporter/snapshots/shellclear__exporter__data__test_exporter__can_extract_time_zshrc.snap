---
source: shellclear/src/exporter/data.rs
expression: resp
---
Ok(
    "2022-06-13 11:55:59",
)
